README.md
# teloscoin-tor
docker container: Telos Coin masternode with Tor and Stubby

Description and purpose:
This repository is a contribution as a resource to easily build a docker image and create a container that contains a hot TELOS Coin wallet configured to run as a masternode, transcendenced uses Tor to connect to the network for privacy and anonymity and all dns queries are encrypted and anonymized using Stubby/Proxychains4/Tor. Tor/Stubby are contained within the same container as Transcendenced making this an all-in-one solution using s6-init to run multiple processes in the same container. The linkage between transcendenced and the local Tor is configured on build of the image. Tor runs as user debian-tor within the container and transcendenced runs as user transcendence. The image is based on ubuntu 18.04 and the Dockerfile will pull this image as a base on build.

<b>BUILD INFO</b><br>
Steps to install:<br>
clone the repository with git:<br>
<code>git clone https://github.com/taperj/teloscoin-tor</code><br>
Change directory to the root of the project:<br>
<code>cd teloscoin-tor</code><br>
Make sure the installer is executable:<br>
<code>chmod +x installer</code><br>
Run the installer<br>
<code>sudo ./install.sh</code><br>

torpassword in transcendence.conf is set to "decentralization" and should be changed to whatever you change the Tor controller password to in the follwing step. Note that you can build and use as is, it has been configured to work but it is highly suggested that you take the time to edit the Dockerfile and transcendence.conf changing the tor control password prior to build, or after build once the container is deployed. I have added the following instructions are in the Dockerfile to guide you:
     
#Hashed Password is "decentralization" change this with tor --hash-password \<yournewpassword\><br>
#and use the ouput to replace the following in /etc/tor/torrc. Make sure to also update transcendence.conf torpassword= with the<br>
#new password in plain text, not hashed.<br>
&&    echo "HashedControlPassword 16:C7F40C06065809EE60D5C0B9086D2BDF88F32495CD1AE06E4571CB8212" >> /etc/tor/torrc \

Make sure to allow port 8051/tcp(or whatever port you specify to run the masternode on in the installer) in your hosts firewall(not within the container), this can usually be accomplished with:<br>
sudo ufw allow 8051/tcp<br>
<br>
<b>Description of encrypted/anonymized DNS linkage:</b><br>
Stubby is a small DNS stub resolver that uses TLS to encrypt its lookups when queried. I have configured Stubby to use proxychains4 which passes the query to a SOCKS5 proxy, being Tor. So any DNS queries within the container are encrypted locally with stubby and the lookup is anonymized over Tor.<br>
<br>
Was this a helpful installer?<br>
feel free to drop me a tip..<br>
BTC: 3HLx5AMe9S5SWzVqLwAib3oyGZm5nAAWKe<br>
TELOS: GPbuPVWKMKBYghszKi8N2iBCJgufmu3Li2<br>
<br>
<br>
<b>Relevant links:</b><br>
<b>S6-INIT:</b> https://skarnet.org/software/s6/ <br>
<b>Tor:</b> https://www.torproject.org/ <br>
<b>Stubby:</b> https://dnsprivacy.org/wiki/display/DP/DNS+Privacy+Daemon+-+Stubby <br>
<b>Proxychains4:</b> https://github.com/rofl0r/proxychains-ng <br>
<b>TELOS Coin:</b> https://teloscoin.org/ <br>
<b>Docker:</b> https://www.docker.com/ <br>

MMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMM<br>
MMMMMMMMMMMMMMMMMWNNNNNNNNNNNNNNNNNNNNNNNNWNNNWWMWWWWWWWWWMMMMMMMMMMMMMMMMM<br>
MMMMMMMMMMMMMMMMKc'',''''''''''''''''''''','',cKNolOO0O00ONMMMMMMMMMMMMMMMM<br>
MMMMMMMMMMMMMMMMO'                           .xKd:lxocod:oXMMMMMMMMMMMMMMMM<br>
MMMMMMMMMMMMMMMMO'                          .xXd:oxxo,ll.cXMMMMMMMMMMMMMMMM<br>
MMMMMMMMMMMMMMMMO'                         .kXo;kkkkxxxd'lXMMMMMMMMMMMMMMMM<br>
MMMMMMMMMMMMMMMMO'                        .kKl:dkkc;dl,ooxNMMMMMMMMMMMMMMMM<br>
MMMMMMMMMMMMMMMMO'                       .kN0:,dxc;:l;;ocoNMMMMMMMMMMMMMMMM<br>
MMMMMMMMMMMMMMMMO'                      'kN0xl,lx;:d;,do,lXMMMMMMMMMMMMMMMM<br>
MMMMMMMMMMMMMMMMO'                     'OXKk;;dxkxxo'okkxxNMMMMMMMMMMMMMMMM<br>
MMMMMMMMMMMMMMMMO'                    'OX0d,:l;:xxOxoxxd,lXMMMMMMMMMMMMMMMM<br>
MMMMMMMMMMMMMMMM0,                  'lOKdkO:oOcc00KKKKKOokNMMMMMMMMMMMMMMMM<br>
MMMMMMMMMMMMMMMMN0OOOOOOOOo.       'OWXo;kNkONxxNMMMMMMMMMMMMMMMMMMMMMMMMMM<br>
MMMMMMMMMMMMMMMMMMMMMMMMMMO'      ,OKkccxk0kkd'lNMMMMMMMMMMMMMMMMMMMMMMMMMM<br>
MMMMMMMMMMMMMMMMMMMMMMMMMMO'     ,O0c,cxl;kOkOodNMMMMMMMMMMMMMMMMMMMMMMMMMM<br>
MMMMMMMMMMMMMMMMMMMMMMMMMMO'    ,0Ko::dOc:dodOodNMMMMMMMMMMMMMMMMMMMMMMMMMM<br>
MMMMMMMMMMMMMMMMMMMMMMMMMMO'   ,0Xl;dOkkkxc'oo'lNMMMMMMMMMMMMMMMMMMMMMMMMMM<br>
MMMMMMMMMMMMMMMMMMMMMMMMMMO'  ;0X0o'ckkkkc'oxkxxNMMMMMMMMMMMMMMMMMMMMMMMMMM<br>
MMMMMMMMMMMMMMMMMMMMMMMMMMO' ;00ldxcokxxd,,l:lxkNMMMMMMMMMMMMMMMMMMMMMMMMMM<br>
MMMMMMMMMMMMMMMMMMMMMMMMMMO,;0Nd,cxdlldxo:;:;ccdNMMMMMMMMMMMMMMMMMMMMMMMMMM<br>
MMMMMMMMMMMMMMMMMMMMMMMMMMKd0K0O;,xl.'okxO:,do'lNMMMMMMMMMMMMMMMMMMMMMMMMMM<br>
MMMMMMMMMMMMMMMMMMMMMMMMMMWNO:lxodd;'cl;;ddc,clxNMMMMMMMMMMMMMMMMMMMMMMMMMM<br>
MMMMMMMMMMMMMMMMMMMMMMMMMMWO;:dxxxd',lolldkdcc;oNMMMMMMMMMMMMMMMMMMMMMMMMMM<br>
MMMMMMMMMMMMMMMMMMMMMMMMMMNo.lxxkxd';ddoxxodxo,lNMMMMMMMMMMMMMMMMMMMMMMMMMM<br>
MMMMMMMMMMMMMMMMMMMMMMMMMMNl.oxkxxl.lOl.co'cdkxkNMMMMMMMMMMMMMMMMMMMMMMMMMM<br>
MMMMMMMMMMMMMMMMMMMMMMMMMMNl.lxxxxc.lOl.cdol,ldkNMMMMMMMMMMMMMMMMMMMMMMMMMM<br>
MMMMMMMMMMMMMMMMMMMMMMMMMMWOdOKKKKOdOXOdOKK0dOKKWMMMMMMMMMMMMMMMMMMMMMMMMMM<br>
MMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMM<br>
MWNNWWWWWWMMMMMMMMMMMMMMWWWMMMWWWMMWNNNWMMMMMMMWNWMMMWNWWMMMMMMMMWWWMMWWWWM<br>


Changelog.md
V. 0.0.1 - Initial Release<br>
<br>
V. 0.0.2 - 12-10-2019<br>
	 - Added custom port specification, can be used for running multiple masternode dockers on one host.<br>
	 - Some formatting changes and code optimization<br>
<br>
V. 0.0.3 - 12-20-2019<br>
         - Added Stubby and Proxychains4 for encrypting and anonymizing DNS queries over Tor<br>
         - Updated Docker base image to Ubuntu 18.04<br>


install.sh
#!/bin/bash
####################################################
# TELOSCOIN MASTERNODE INSTALLER WITH TOR          #
#                                                  #
#  https://github.com/taperj/teloscoin-tor         #
#                                                  #
#  V. 0.0.3                                        #
#                                                  #
#  By: taperj                                      #
#                                                  #
####################################################
RED='\033[0;31m'
NC='\033[0m'
YELLOW='\033[1;33m'
GREEN='\033[0;32m'
WHITE='\033[1;37m'
#
#
#Root user check
if [ "$EUID" -ne 0 ]
  then printf "${YELLOW}Please run as root or with sudo. ${RED}Exiting.${NC}\n"
  exit
fi
#
#
#printf "${YELLOW}${NC}\n"
printf "${YELLOW}MMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMM
MMMMMMMMMMMMMMMMMWNNNNNNNNNNNNNNNNNNNNNNNNWNNNWWMWWWWWWWWWMMMMMMMMMMMMMMMMM
MMMMMMMMMMMMMMMMKc'',''''''''''''''''''''','',cKNolOO0O00ONMMMMMMMMMMMMMMMM
MMMMMMMMMMMMMMMMO'                           .xKd:lxocod:oXMMMMMMMMMMMMMMMM
MMMMMMMMMMMMMMMMO'                          .xXd:oxxo,ll.cXMMMMMMMMMMMMMMMM
MMMMMMMMMMMMMMMMO'                         .kXo;kkkkxxxd'lXMMMMMMMMMMMMMMMM
MMMMMMMMMMMMMMMMO'                        .kKl:dkkc;dl,ooxNMMMMMMMMMMMMMMMM
MMMMMMMMMMMMMMMMO'                       .kN0:,dxc;:l;;ocoNMMMMMMMMMMMMMMMM
MMMMMMMMMMMMMMMMO'                      'kN0xl,lx;:d;,do,lXMMMMMMMMMMMMMMMM
MMMMMMMMMMMMMMMMO'                     'OXKk;;dxkxxo'okkxxNMMMMMMMMMMMMMMMM
MMMMMMMMMMMMMMMMO'                    'OX0d,:l;:xxOxoxxd,lXMMMMMMMMMMMMMMMM
MMMMMMMMMMMMMMMM0,                  'lOKdkO:oOcc00KKKKKOokNMMMMMMMMMMMMMMMM
MMMMMMMMMMMMMMMMN0OOOOOOOOo.       'OWXo;kNkONxxNMMMMMMMMMMMMMMMMMMMMMMMMMM
MMMMMMMMMMMMMMMMMMMMMMMMMMO'      ,OKkccxk0kkd'lNMMMMMMMMMMMMMMMMMMMMMMMMMM
MMMMMMMMMMMMMMMMMMMMMMMMMMO'     ,O0c,cxl;kOkOodNMMMMMMMMMMMMMMMMMMMMMMMMMM
MMMMMMMMMMMMMMMMMMMMMMMMMMO'    ,0Ko::dOc:dodOodNMMMMMMMMMMMMMMMMMMMMMMMMMM
MMMMMMMMMMMMMMMMMMMMMMMMMMO'   ,0Xl;dOkkkxc'oo'lNMMMMMMMMMMMMMMMMMMMMMMMMMM
MMMMMMMMMMMMMMMMMMMMMMMMMMO'  ;0X0o'ckkkkc'oxkxxNMMMMMMMMMMMMMMMMMMMMMMMMMM
MMMMMMMMMMMMMMMMMMMMMMMMMMO' ;00ldxcokxxd,,l:lxkNMMMMMMMMMMMMMMMMMMMMMMMMMM
MMMMMMMMMMMMMMMMMMMMMMMMMMO,;0Nd,cxdlldxo:;:;ccdNMMMMMMMMMMMMMMMMMMMMMMMMMM
MMMMMMMMMMMMMMMMMMMMMMMMMMKd0K0O;,xl.'okxO:,do'lNMMMMMMMMMMMMMMMMMMMMMMMMMM
MMMMMMMMMMMMMMMMMMMMMMMMMMWNO:lxodd;'cl;;ddc,clxNMMMMMMMMMMMMMMMMMMMMMMMMMM
MMMMMMMMMMMMMMMMMMMMMMMMMMWO;:dxxxd',lolldkdcc;oNMMMMMMMMMMMMMMMMMMMMMMMMMM
MMMMMMMMMMMMMMMMMMMMMMMMMMNo.lxxkxd';ddoxxodxo,lNMMMMMMMMMMMMMMMMMMMMMMMMMM
MMMMMMMMMMMMMMMMMMMMMMMMMMNl.oxkxxl.lOl.co'cdkxkNMMMMMMMMMMMMMMMMMMMMMMMMMM
MMMMMMMMMMMMMMMMMMMMMMMMMMNl.lxxxxc.lOl.cdol,ldkNMMMMMMMMMMMMMMMMMMMMMMMMMM
MMMMMMMMMMMMMMMMMMMMMMMMMMWOdOKKKKOdOXOdOKK0dOKKWMMMMMMMMMMMMMMMMMMMMMMMMMM
MMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMM
MWNNWWWWWWMMMMMMMMMMMMMMWWWMMMWWWMMWNNNWMMMMMMMWNWMMMWNWWMMMMMMMMWWWMMWWWWM${NC}\n"
printf "${RED}***************************************************************************${NC}\n"
printf "${RED}****************${YELLOW}WELCOME TO THE TELOSCOIN MASTERNODE INSTALLER${RED}****************${NC}\n"
printf "${YELLOW}******THIS SCRIPT WILL INSTALL A DOCKER CONTAINER WITH TRANSCENDENCED AND TOR*****${NC}\n"
printf "${RED}***************************************************************************${NC}\n"
#
#
#Get install info:
#
printf "${WHITE}Enter the masternode privkey and hit enter:${NC}\n"
read MASTERNODEPRIVKEY
#
#
#Get masternode's public ip
#
printf "${WHITE}Detecting Public IP...${NC}\n"
#
#Check for cURL
if ! [ -x "$(command -v curl)" ]; then
        printf "${RED}cURL is not detected or not executable.${GREEN} Installing cURL.${NC}\n"
        apt-get -y install curl
fi
#
PUBLICIP=$(curl -s ifconfig.me)
printf "${GREEN}Public IP detected is: $PUBLICIP${NC}\n"
#
printf "${WHITE}Enter the ip you would like to use for the masternode and hit enter:${NC}\n"
read MASTERNODEADDR
#
#
#Port specification
#make sure not to conflict with tor on 9050 and 9051
#
printf "${WHITE}Enter the port number you'd like transcendenced to listen on, default Port 8051 will be used if no port specified.${NC}\n"
read PORTNUMBER

if [ "$PORTNUMBER" != "" ]
  then
        if [ "$PORTNUMBER" = "9050" ] || [ "$PORTNUMBER" = "9051" ]
                then
                        printf "${RED}Port $PORTNUMBER specified in user input. $PORTNUMBER is reserved for Tor. Exiting.${NC}\n"
                        printf "${RED}PLEASE RE-RUN THE SCRIPT SELECTING A DIFFERENT PORT.${NC}\n"
                        exit
        fi
          printf "${YELLOW}Port $PORTNUMBER specified in user input. Port $PORTNUMBER will be configured.${NC}\n"
  else
          printf "${YELLOW}No port number specified. Default Port 8051 will be used.${NC}\n"
          PORTNUMBER=8051
fi

#
#
#RPC
#
printf "${WHITE}Enter a username for RPC:${NC}\n"
read RPCUSER
printf "${WHITE}Enter a password for RPC:${NC}\n"
read RPCPASSWORD
printf "${WHITE}SANITY CHECK...${NC}\n"
#
#
#Sanity check
#
#####
#
#Check for docker
#
if ! [ -x "$(command -v docker)" ]; then
	printf "${RED}docker is not detected or not executable.${GREEN} Installing docker.${NC}\n"
	apt-get -y install docker docker.io
fi

if [ -x "$(command -v docker)" ]; then
        printf "${YELLOW}docker detected and executable.${GREEN} Continuing.${NC}\n"
fi

#
#
#Check for files and dirs needed
for file in transcendence.conf Dockerfile services/transcendenced/run services/transcendenced/finish services/tor/run services/tor/finish
do
if [ ! -f $file ]; then
	printf "${RED}SANITY CHECK FAILED: $file not found in the current directory, exiting!${NC}\n"
	exit
fi
done
#
#
##
for dir in services services/transcendenced services/tor 
do
if [ ! -d $dir ]; then
	printf "${RED}SANITY CHECK FAILED: $dir directory not found, exiting!${NC}\n"
	exit
	fi
done
##
printf "${GREEN}SANITY CHECK PASSED!${NC}\n"
printf "${YELLOW}BEGINNING INSTALL...${NC}\n"
#
#
#
#
#Edit transcendence.conf:
#
printf "${YELLOW}Editing transcendence.conf...${NC}\n"
sed -i "s/masternodeprivkey=/masternodeprivkey=$MASTERNODEPRIVKEY/g" transcendence.conf
sed -i "s/masternodeaddr=/masternodeaddr=$MASTERNODEADDR/g" transcendence.conf
sed -i "s/rpcuser=/rpcuser=$RPCUSER/g" transcendence.conf
sed -i "s/rpcpassword=/rpcpassword=$RPCPASSWORD/g" transcendence.conf
sed -i "s/^port=/port=$PORTNUMBER/g" transcendence.conf
#
#
#Edit Dockerfile
printf "${YELLOW}Editing Dockerfile...${NC}\n"
sed -i "s/HiddenServicePort 8051 127.0.0.1:8051/HiddenServicePort $PORTNUMBER 127.0.0.1:$PORTNUMBER/g" Dockerfile
#
#
#Build image
#
printf "${YELLOW}Building docker image telos-tor...${NC}\n"
docker build -t telos-tor-$PORTNUMBER .
#
#Create container
#
printf "${YELLOW}Creating container telos-tor...${NC}\n"
docker create --name telos-tor-$PORTNUMBER --restart=always -p $PORTNUMBER:$PORTNUMBER telos-tor-$PORTNUMBER:latest
#
#
#Start container
#
printf "${YELLOW}Starting container telos-tor...${NC}\n"
docker container start telos-tor-$PORTNUMBER
sleep 10
docker ps
printf "${GREEN}INSTALLATION COMPLETE.${NC}\n"
printf "${YELLOW}ONCE SYNCED YOU CAN GET THE TOR(onion) ADDRESS TO ADD TO YOUR COLD WALLET masternode.conf as server address with:${NC}\n"
printf "${WHITE}$ sudo docker container exec telos-tor-$PORTNUMBER grep AddLocal /home/transcendence/.transcendence/debug.log${NC}\n"
printf "${YELLOW}THE ABOVE COMMAND SHOULD OUTPUT SOMETHING LIKE THIS EXAMPLE OUTPUT:${NC}\n"
printf "${WHITE}2019-11-24 02:33:16 AddLocal(zsddfken27kdsdx.onion:$PORTNUMBER,4)${NC}\n"
printf "${YELLOW}in this example you would add ${GREEN}zsddfken27kdsdx.onion:$PORTNUMBER${YELLOW} to your cold wallet masternode.conf as ip addr for this alias. Yours will be different than the example.${NC}\n"
printf "${RED}IMPORTANT: IF YOU ARE RUNNING A FIREWALL MAKE SURE TO ALLOW PORT $PORTNUMBER/TCP FOR transcendenced${NC}\n"
printf "${YELLOW}Tips for the developer:${NC}\n"
printf "${YELLOW}BTC: 3HLx5AMe9S5SWzVqLwAib3oyGZm5nAAWKe${NC}\n"
printf "${YELLOW}TELOS: GPbuPVWKMKBYghszKi8N2iBCJgufmu3Li2${NC}\n"
printf "${YELLOW}HAVE FUN ANONYMOUS TELOSCOIN MASTERNODING!${NC}\n"
